### PR TITLE
[Peertube] Add Channel extractor

### PR DESCRIPTION
**Pull request by u-spec-png**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fix

### CODE REVIEW

**Comment by pukkandan** [comment]
Cant u create a baseextractor and generalize the code with playlistIE?

File: yt_dlp/extractor/peertube.py [hunk 1]
```diff
@@ -1343,7 +1343,7 @@ def _fetch_page(self, host, uuid, page):
         page += 1
         video_data = self._call_api(
             host, uuid, f'/videos?sort=-createdAt&start={self._PAGE_SIZE * (page - 1)}&count={self._PAGE_SIZE}',
-            note=f'Downloading page {page}').get('data', [])
+            note=f'Downloading page {page}').get('data', {})
```
Comment (Ashish0804) [inline yt_dlp/extractor/peertube.py:1346]: Why change this to dict?

File: yt_dlp/extractor/peertube.py [hunk 2]
```diff
@@ -1369,3 +1369,60 @@ def _real_extract(self, url):
...
+    def _call_api(self, host, name, path, base, note=None, **kwargs):
+        return self._download_json(
+            self._API_BASE % (host, base, name, path), name,
+            note=note, **kwargs)
+
+    def _fetch_page(self, host, uuid, type, page):
+        page += 1
+        video_data = self._call_api(
+            host, uuid, f'/videos?sort=-createdAt&start={self._PAGE_SIZE * (page - 1)}&count={self._PAGE_SIZE}&nsfw=both',
+            type, note=f'Downloading page {page}').get('data', {})
```
Comment (Ashish0804) [inline yt_dlp/extractor/peertube.py:1406]: here aswell, it should be list and not dict
doing for each on dict gives you the keys.

**Comment by u-spec-png** [comment]
> Cant u create a baseextractor and generalize the code with playlistIE?

Done

File: yt_dlp/extractor/peertube.py [hunk 3]
```diff
@@ -1295,15 +1295,37 @@ def channel_data(field, type_):
...
+            type, note=f'Downloading page {page}').get('data', [])
+        for video in video_data:
+            shortUUID = video.get('shortUUID') or try_get(video, lambda x: x['video']['shortUUID'])
+            video_title = video.get('name') or try_get(video, lambda x: x['video']['name'])
+            yield self.url_result(
+                f'https://{host}/w/{shortUUID}', PeerTubeIE.ie_key(),
+                video_id=shortUUID, video_title=video_title)
+
+
+class PeerTubePlaylistIE(PeerTubeBaseIE, InfoExtractor):
```
Comment (pukkandan) [inline yt_dlp/extractor/peertube.py:1321]:
```suggestion
class PeerTubePlaylistIE(PeerTubeBaseIE):
```
same below

File: yt_dlp/extractor/peertube.py [hunk 4]
```diff
@@ -1332,28 +1354,10 @@ class PeerTubePlaylistIE(InfoExtractor):
...
-            host, uuid, f'/videos?sort=-createdAt&start={self._PAGE_SIZE * (page - 1)}&count={self._PAGE_SIZE}',
-            note=f'Downloading page {page}').get('data', {})
-        for video in video_data:
-            shortUUID = try_get(video, lambda x: x['video']['shortUUID'])
-            video_title = try_get(video, lambda x: x['video']['name'])
-            yield self.url_result(
-                f'https://{host}/w/{shortUUID}', PeerTubeIE.ie_key(),
-                video_id=shortUUID, video_title=video_title)
 
     def _real_extract(self, url):
```
Comment (pukkandan) [inline yt_dlp/extractor/peertube.py:1358]: Seems same as channelIE except type is hardcoded to `video-playlists`

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/peertube.py (hunk 4 above)
Comment (u-spec-png) [reply yt_dlp/extractor/peertube.py:1358]: Should I merge it into one extractor?

Comment (pukkandan) [reply yt_dlp/extractor/peertube.py:1358]: I was thinking of splitting it out into a function (say `_extract_playlist`) that takes `host, type, id`. But if merging is easier, then even better

Comment (u-spec-png) [reply yt_dlp/extractor/peertube.py]: https://github.com/yt-dlp/yt-dlp/pull/1023/commits/58402809978be8183b7da022f7d349e8dcb99e08 is this good?

File: yt_dlp/extractor/peertube.py [hunk 5]
```diff
@@ -1295,36 +1295,11 @@ def channel_data(field, type_):
...
-class PeerTubePlaylistIE(PeerTubeBaseIE):
+class PeerTubePlaylistIE(InfoExtractor):
     IE_NAME = 'PeerTube:Playlist'
     _VALID_URL = r'''(?x)
-                    (?:
-                        https?://(?P<host>%s)/w/p/
-                    )
-                    (?P<id>%s)
+                        https?://(?P<host>%s)/(?P<type>(?:a|c|w/p))/
+                    (?P<id>(?:%s|chocobozzz))
```
Comment (pukkandan) [inline yt_dlp/extractor/peertube.py:1302]: ?

File: yt_dlp/extractor/peertube.py [hunk 6]
```diff
@@ -1395,19 +1345,51 @@ class PeerTubeChannelIE(PeerTubeBaseIE, InfoExtractor):
...
+    _PAGE_SIZE = 30
 
-    def _real_extract(self, url):
-        type, host, channel_id = self._match_valid_url(url).group('type', 'host', 'id')
-        type = 'accounts' if type == 'a' else 'video-channels'
-        channel_info = self.call_api(host, channel_id, '', type, note='Downloading channel information', fatal=False)
+    def call_api(self, host, name, path, base, note=None, **kwargs):
+        return self._download_json(
+            self._API_BASE % (host, base, name, path), name,
+            note=note, **kwargs)
```
Comment (pukkandan) [inline yt_dlp/extractor/peertube.py:1354]: note can also be in kwargs, no?

File: yt_dlp/extractor/peertube.py [hunk 7]
```diff
@@ -1395,19 +1345,51 @@ class PeerTubeChannelIE(PeerTubeBaseIE, InfoExtractor):
...
+            timestamp=playlist_timestamp, channel=channel, channel_id=channel_id, thumbnail=thumbnail)
+
+    def _real_extract(self, url):
+        type, host, id = self._match_valid_url(url).group('type', 'host', 'id')
+        if type == 'a':
+            type = 'accounts'
+        elif type == 'c':
+            type = 'video-channels'
+        elif type == 'w/p':
+            type = 'video-playlists'
```
Comment (pukkandan) [inline yt_dlp/extractor/peertube.py:1394]: you can add this in the class definition
```py
    _TYPES = {
        'a': 'accounts',
        ...
    }
```
and then do
```py
       type = self._TYPES[type]
```

Then u can also use `'|'.join(_TYPES.keys())` in `_VALID_URL`

**Review by pukkandan — APPROVED** [review]
Everything else looks good

**Pull request merged**